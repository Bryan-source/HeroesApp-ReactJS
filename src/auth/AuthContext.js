

/* En concreto, un componente de orden superior es una función que recibe un componente y devuelve un nuevo componente.
El nombre lleva mayúsculas por lo mismo de que es un higherOrderComponent
const EnhancedComponent = higherOrderComponent(WrappedComponent);
Mientras que un componente transforma props en interfaz de usuario, un componente de orden superior transforma un componente en otro. */

/* Este componente no necesita ser probado debido a que esto ya es cuestión de React en sí */
import { createContext } from "react";

export const AuthContext = createContext();