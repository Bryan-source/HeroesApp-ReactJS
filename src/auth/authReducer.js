//Autentificar cuando haces login

import { types } from "../types/types";

export const authReducer = (state = [], action) => {
    switch (action.type) {
        case types.login:
            return {
                /* ...action.payload se puede mandar el nombre del usuario el email etc cuando se haga login. */
                ...action.payload,
                logged: true
            }
            
        case types.logout:
            return {
                logged: false
            }
        default:
            return state;
    }
}