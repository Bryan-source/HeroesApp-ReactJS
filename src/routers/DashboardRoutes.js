import React from 'react'
import { Navbar } from '../components/ui/Navbar'
import {
    Switch,
    Route,
    Redirect
  } from "react-router-dom";
import { HeroScreen } from '../components/heroes/HeroScreen';
import { MarvelScreen } from '../components/marvel/MarvelScreen';
import { DcScreen } from '../components/dc/DcScreen';
import { SearchScreen } from '../components/search/SearchScreen';

//La diferencia entre con el AppRouter es que este no tiene router solamente tiene las rutas y el navbar
export const DashboardRoutes = () => {
    return (
        <>
        {/* En este punto se observa que el Navbar no se encuentra en una ruta es decir dentro de un (Route) por lo tanto no hereda las props que el route otorga a los elementos que si est√°n dentro de el.*/}
            <Navbar/>

            <div className = 'container mt-2'>
                <Switch>
                    <Route exact path="/marvel" component = {MarvelScreen}/>                    
                    <Route exact path="/hero/:heroId" component = {HeroScreen}/>
                    <Route exact path="/dc" component = {DcScreen}/>
                    <Route exact path="/search" component = {SearchScreen}/>

                    <Redirect to = "/marvel"/>                                        
                </Switch>
            </div>
        </>
    )
}
