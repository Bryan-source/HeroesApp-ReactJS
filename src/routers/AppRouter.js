import React, { useContext } from 'react'
import {
    BrowserRouter as Router,
    Switch,
  } from "react-router-dom";
import { AuthContext } from '../auth/AuthContext';
import { LoginScreen } from '../components/login/LoginScreen';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {

    //useContext donde se definen user y dispatch
    const {user} = useContext(AuthContext);

    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute 
                        exact 
                        path="/login" 
                        component = {LoginScreen}
                        isAuthenticated = {user.logged}
                    />
                        
                    {/* El path es parte del ...rest de PrivateRoute */}
                    {/* En el localStorage guarda el pathname del rest de la respectiva ruta del DashBoard */}
                    <PrivateRoute 
                        path="/" 
                        component = {DashboardRoutes}
                        isAuthenticated = {user.logged}   
                    />                                    
                </Switch>
            </div>
        </Router>
    )
}
