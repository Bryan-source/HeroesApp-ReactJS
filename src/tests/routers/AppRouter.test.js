import React from 'react';
import {mount} from 'enzyme'
import { AuthContext } from '../../auth/AuthContext';
import { AppRouter } from '../../routers/AppRouter';

describe('Testing AppRouter.js', () => {

    const context = {
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    }
    test('should show login if it is not authenticated', () => {
        const wrapper = mount(
            <AuthContext.Provider value = {context}>
                <AppRouter/>
            </AuthContext.Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });

    test('should show marvel if it is authenticated', () => {
        const context = {
            dispatch: jest.fn(),
            user: {
                logged: true,
                user: 'Bryan'
            }
        }

        const wrapper = mount(
            <AuthContext.Provider value = {context}>
                <AppRouter/>
            </AuthContext.Provider>
        );
        expect(wrapper.find('.navbar').exists()).toBe(true)

    });
    
})
