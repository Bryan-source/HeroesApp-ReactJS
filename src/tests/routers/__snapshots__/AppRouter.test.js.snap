// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing AppRouter.js should show login if it is not authenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login
    </h1>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <input
          className="form-control"
          id="exampleInputText1"
          name="userName"
          onChange={[Function]}
          placeholder="Write your name"
          type="text"
          value=""
        />
      </div>
      <button
        className="btn btn-primary"
        type="submit"
      >
         Login 
      </button>
    </form>
  </div>
</div>
`;
